Assets/Scripts/EOSTransferTask.cs

```diff
@@ -0,0 +1,305 @@
+        public delegate void OnProgressUpdatedDelegate(float progress);
+
+        /// <summary>
+        /// An optional task that can be utilized to inform the user of file transfer progress.
+        /// </summary>
+        public event OnProgressUpdatedDelegate OnProgressUpdated;
+
+        /// <summary>
+        /// If this task is a download task, this is where the data will be put when the task is complete.
+        /// If this is an upload task, this is where the data should be put.
+        /// </summary>
+        public byte[] Data { get; protected set; } = new byte[] { };
+
+        /// <summary>
+        /// If this task has run to completion, or encountered an error, this ResultCode should be set.
+        /// You can determine that an EOSTransferTask is complete if it has any result code.
+        /// </summary>
+        public Result? ResultCode;
+
+        /// <summary>
+        /// Percentage indicator of the work's completeness.
+        /// Either read from this value or subscribe to <see cref="OnProgressUpdated"/> to receive updates on progress amounts.
+        /// </summary>
+        public float Progress;
+
+        /// <summary>
+        /// Inner backing for the total size of the transfer.
+        /// </summary>
+        private uint transferSize = 0;
+
+        /// <summary>
+        /// Indexes the current byte count that has been processed.
+        /// </summary>
+        private uint currentIndex = 0;
+
+        /// <summary>
+        /// The total size of the transfer operation in bytes.
+        /// </summary>
+        public uint TotalSize
+        {
+            get
+            {
+                return transferSize;
+            }
+            private set
+            {
+                transferSize = value;
+
+                if (transferSize > FILE_MAX_SIZE_BYTES)
+                {
+                    Debug.LogError("[EOS SDK] Player data storage: data transfer size exceeds max file size.");
+                    transferSize = FILE_MAX_SIZE_BYTES;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Constructor for download scenarios.
+        /// </summary>
+        /// <param name="innerTask">A task completion source that can be used to indicate when the task is completed.</param>
+        public EOSTransferTask(TaskCompletionSource<byte[]> innerTask)
+        {
+            InnerTaskCompletionSource = innerTask;
+        }
+
+        /// <summary>
+        /// Constructor for upload scenarios.
+        /// </summary>
+        /// <param name="innerTask">A task completion source that can be used to indicate when the task is completed.</param>
+        /// <param name="data">The data to upload, in byte[] form.</param>
+        public EOSTransferTask(TaskCompletionSource<byte[]> innerTask, byte[] data)
+        {
+            InnerTaskCompletionSource = innerTask;
+            Data = data;
+            TotalSize = (uint)data.Length;
+        }
```

Comment:
Since the intent behind `EOSTransferTask` is to have it be agnostic to whether the operation is an upload or a download, the difference in implementation between those two things should be reflected in the deriving classes, not in the base implementation. 

`EOSTransferTask` should contain functionality that is common to both scenarios and exclude any implementation that is unique to any.